File: solutions/java/1448.java
/*
 * Leetcode 1448: Count Good Nodes in Binary Tree. Java. Medium.
 * O(n) runtime and O(n) space complexity.
 * Recursive depth-first search traversal keeping track of the minimum allowable value, 
 * used to count the number of good nodes in the tree.
 * 5/3/2024 Winston Tsui
*/

class Solution {
    public int goodNodes(TreeNode root) {
        return goodNodes(root, Integer.MIN_VALUE);
    }

    private int goodNodes(TreeNode root, int min) {
        if (root == null)
            return 0;
        if (root.val < min)
            return goodNodes(root.left, Math.max(root.val, min)) + goodNodes(root.right, Math.max(root.val, min));
        return 1 + goodNodes(root.left, Math.max(root.val, min)) + goodNodes(root.right, Math.max(root.val, min));
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: solutions/python/1448.py
"""
Leetcode 1448: Count Good Nodes in Binary Tree. Python. Medium.
O(N) runtime and O(H) space complexity, where N is the number of nodes and H is the height of the tree.
DFS to traverse the tree while carrying the max value seen so far along the path.
3/30/2025 Winston Tsui
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def numGoodNodes(root: TreeNode, maxSoFar: int) -> int:
            if not root:
                return 0
            count = 1 if root.val >= maxSoFar else 0
            maxSoFar = max(maxSoFar, root.val)
            count += numGoodNodes(root.left, maxSoFar)
            count += numGoodNodes(root.right, maxSoFar)
            return count

        return numGoodNodes(root, root.val)
